pkg/watcher/reconciler/dynamic/dynamic.go

```diff
@@ -314,101 +319,141 @@ func (r *Reconciler) sendLog(ctx context.Context, o results.Object) error {
+	podLogs, err := req.Stream(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer podLogs.Close()
+
+	if err != nil && errors.IsNotFound(err) {
+		msg := fmt.Sprintf("error getting logs for pod %s container %s: %s", pod, container, err.Error())
+		msgBytes := []byte(msg)
+		return msgBytes, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	buf := new(bytes.Buffer)
+	_, err = io.Copy(buf, podLogs)
+	return buf.Bytes(), err
+}
+
+func (r *Reconciler) streamLogs(ctx context.Context, o results.Object, labelKey, logName string) error {
 	logger := logging.FromContext(ctx)
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	logger.Debugw("Streaming log started",
+		zap.String("namespace", o.GetNamespace()),
+		zap.String("kind", o.GetObjectKind().GroupVersionKind().Kind),
+		zap.String("name", o.GetName()),
+	)
 	logsClient, err := r.resultsClient.UpdateLog(ctx)
 	if err != nil {
 		return fmt.Errorf("failed to create UpdateLog client: %w", err)
 	}
 
 	writer := logs.NewBufferedWriter(logsClient, logName, logs.DefaultBufferSize)
 
-	tknParams := &cli.TektonParams{}
-	tknParams.SetNamespace(o.GetNamespace())
-	// KLUGE: tkn reader.Read() will raise an error if a step in the TaskRun failed and there is no
-	// Err writer in the Stream object. This will result in some "error" messages being written to
-	// the log.
-
-	reader, err := tknlog.NewReader(logType, &tknopts.LogOptions{
-		AllSteps:        true,
-		Params:          tknParams,
-		PipelineRunName: o.GetName(),
-		TaskrunName:     o.GetName(),
-		Stream: &cli.Stream{
-			Out: writer,
-			Err: writer,
-		},
-	})
-	if err != nil {
-		return fmt.Errorf("failed to create tkn reader: %w", err)
+	inMemWriteBufferStdout := bytes.NewBuffer(make([]byte, 0))
+
+	lo := metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", labelKey, o.GetName()),
 	}
-	logChan, errChan, err := reader.Read()
+	var pods *corev1.PodList
+	pods, err = r.kubernetesClientset.CoreV1().Pods(o.GetNamespace()).List(ctx, lo)
 	if err != nil {
-		return fmt.Errorf("error reading from tkn reader: %w", err)
+		return err
 	}
 
-	errChanRepeater := make(chan error)
-	go func(echan <-chan error, o metav1.Object) {
-		writeErr := <-echan
-		errChanRepeater <- writeErr
-
-		_, err := writer.Flush()
-		if err != nil {
-			logger.Error(err)
-		}
-		if err = logsClient.CloseSend(); err != nil {
-			logger.Error(err)
+	for _, pod := range pods.Items {
+		// fyi gocritic complained about employing 'containers := append(pod.Spec.InitContainers, pod.Spec.Containers...)'
+		containers := []corev1.Container{}
+		copy(containers, pod.Spec.InitContainers)
+		containers = append(containers, pod.Spec.Containers...)
+		for _, container := range containers {
+			ba, podLogsErr := r.getPodLogs(ctx, o.GetNamespace(), pod.Name, container.Name)
+			if podLogsErr != nil {
+				return podLogsErr
+			}
+			hdr := fmt.Sprintf("*** Logs for pod %s container %s ***\n", pod.Name, container.Name)
```

Comment:
Shoudl we keep the format same as the tkn logs?
https://github.com/tektoncd/cli/blob/0369f038b9fa32fe746eb1b32db6345822a9dc5c/pkg/log/writer.go#L56-L61

Reply:
+1

Reply:
as long as we merge https://github.com/tektoncd/results/pull/712 first so we can get some immediately relief with respect to the memory leak, then yes, I'll look more into the "Rainbow" decorator or whatever it was that I saw while investigating things, and see about keeping that format.

Also, I want to get a baseline in our prod env (vs. our separate stress tests in test envs) with the current multi-threaded format before moving on to this PR's further reduction in threads.

Reply:
I did not notice that the colouring package is used. Since we are storing the logs I don't think we should do the "rainbow" colours, but the format with [task-name] and [step-name] is what I meant.

Reply:
switch to old format is up and passed e2e's (though a second e2e run is in progress as I type as I've removed debug and squashed related commits)

PTAL @sayan-biswas @avinal